### DISCUSSION TITLE
Problem with websocket in writing plugin

### DISCUSSION THREAD

**Discussion by Dyupich** [streamlink/streamlink#4655] [Q&A] [answered]
Hello! I want to write plugin for site using websocket. 
For example I took plugin: src/streamlink/plugins/ustreamtv.
But I have some problem. When I trying to use wsclient.start() (wsclient is instance of a streamlink.plugin.api.websocket.WebsocketClient) I got this error:

Found matching plugin myplugin for URL nameofmylink
error: No playable streams found on this URL: nameofmylink

When I use wsclient.run() method and interrupt program with CTRL+C I got this:

[cli][info] Found matching plugin myplugin for URL nameofmylink
[plugin.api.websocket][error] <------------------- My Ctrl+C Interruption
[cli][info] Available streams: mystream (worst, best)
[cli][info] Opening stream: mystream (stream)
[cli][info] Starting player: "pathToPlayer"
Traceback (most recent call last):
  File "...\venv\Scripts\streamlink-script.py", line 33, in <module>
    sys.exit(load_entry_point('streamlink', 'console_scripts', 'streamlink')())
  File "...\streamlink\src\streamlink_cli\main.py", line 1101, in main
    handle_url()
  File "...\streamlink\src\streamlink_cli\main.py", line 616, in handle_url
    handle_stream(plugin, streams, stream_name)
  File "...\streamlink\src\streamlink_cli\main.py", line 484, in handle_stream
    success = output_stream(stream, formatter)
  File "...\streamlink\src\streamlink_cli\main.py", line 344, in output_stream
    stream_fd, prebuffer = open_stream(stream)
  File "...\streamlink\src\streamlink_cli\main.py", line 322, in open_stream
    prebuffer = stream_fd.read(8192)
AttributeError: 'NoneType' object has no attribute 'read'

How can I use Websocket to dodge this errors and make my own plugin?

**Comment by bastimeyer** [comment, marked as answer]
Whatever you're trying to do, your custom plugin is not returning a `Stream` class instance, hence the `AttributeError: 'NoneType' object has no attribute 'read'` error in the `streamlink_cli.main` module.

However, you have provided exactly zero information about your plugin implementation, so nobody will be able to help.

The `WebsocketClient` is a utility class which runs a websocket connection in a separate thread. `KeyboardInterrupt`s only get raised on the main thread, which means you'll have to handle this appropriately.
